# Oczyszczanie nieuporzadkowanych danych
___
W tym rozdziale bede badac dane za pomoca **eksploracyjnej analizy danych** - EDA.

In [14]:
import pandas as pd

data = pd.read_csv("data/csv/employee.csv")

data.head() # sprawdzam 5 pierwszych rekordow

name   age   income gender  department grade  performance_score
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723
1        S Kumar   NaN  16000.0      F     Finance    G0                520
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674
3      Ying Chin  45.0  65000.0      F       Sales    G3                556
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711

In [5]:
data.tail() # sprawdzam 5 ostatnich rekordow

name   age   income gender  department grade  performance_score
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711
5  Satyam Sharma   NaN  62000.0    NaN       Sales    G3                649
6   James Authur  54.0      NaN      F  Operations    G3                 53
7     Josh Wills  54.0  52000.0      F     Finance    G3                901
8       Leo Duck  23.0  98000.0      M       Sales    G4                709

In [6]:
# spawdzam liste kolumn
data.columns

Index(['name', 'age', 'income', 'gender', 'department', 'grade',
       'performance_score'],
      dtype='object')

In [11]:
#sprawdzam wymiary obiektu DataFrame
data.shape # zawiera 9 rzedow, 7 kolumn

(9, 7)

In [15]:
#wyswietlam informacje o obiekcie DataFrame
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   name               9 non-null      object 
 1   age                7 non-null      float64
 2   income             7 non-null      float64
 3   gender             7 non-null      object 
 4   department         9 non-null      object 
 5   grade              9 non-null      object 
 6   performance_score  9 non-null      int64  
dtypes: float64(2), int64(1), object(4)
memory usage: 632.0+ bytes


##### w powyszym listingu widze ze 4 kolumny zawieraja obiekty, 2 kolumny zawiera float64 czyli wartosci zmiennoprzecinkowe, i jedna kolumna zawiera warosc int64 , czyli staloprzecinkowa. W age, income, gender - brakuje wartosci (7 non -null), zamoias (9 non-null)

In [17]:
# wyswietlam statystyke opisowa DataFremu
data.describe()

age        income  performance_score
count   7.000000      7.000000           9.000000
mean   40.428571  52857.142857         610.666667
std    12.204605  26028.372797         235.671912
min    23.000000  16000.000000          53.000000
25%    31.000000  38500.000000         556.000000
50%    45.000000  52000.000000         674.000000
75%    49.500000  63500.000000         711.000000
max    54.000000  98000.000000         901.000000

##### Za pomoca powyzszego listingu moge stwierdzić , ze przedzlial wieku pracownikow jest w przedzial od 23 do 54 lat, srednia wieku wynosi 40 lat, a mediana 45 lat, srednii zarobki wynosza 53 000. A najwiezka skutecznosc pracownikow wynosla 901

### Filtrowanie danych w celu pozbycia sie szumu
___
**Dane moga byc filtowane po kolumnach i po rzedach**

##### Filtorowanie po kolumnach:
za pomoca metody `filter()`, w funkcje przekazuje kolumne odnoscie jakich chce pofiltowac dane. Lub za pomoca fragmentowanie , i listy Pythona , przekazuje wiele kolumn z obiektu DataFrame i otrzymuje filtowanie po kolumnach bez uzyciu funkcji `filter()`

In [44]:
data[['name','age']]

name   age
0    Allen Smith  45.0
1        S Kumar   NaN
2    Jack Morgan  32.0
3      Ying Chin  45.0
4  Dheeraj Patel  30.0
5  Satyam Sharma   NaN
6   James Authur  54.0
7     Josh Wills  54.0
8       Leo Duck  23.0

In [24]:
data.filter(['name','age'])

name   age
0    Allen Smith  45.0
1        S Kumar   NaN
2    Jack Morgan  32.0
3      Ying Chin  45.0
4  Dheeraj Patel  30.0
5  Satyam Sharma   NaN
6   James Authur  54.0
7     Josh Wills  54.0
8       Leo Duck  23.0

Na podobej zaszadzie moge filtrowac kolumny przy uzyciu fragmentowanie, w takim przypadku przekazuje jedna kolumne, i w rezultacie otrzymuje obiekt Series, lub przekazuje jedna kolumne za pomoca listy Pythona:

In [28]:
data['name'] # otrzymuje obiekt Series

0      Allen Smith
1          S Kumar
2      Jack Morgan
3        Ying Chin
4    Dheeraj Patel
5    Satyam Sharma
6     James Authur
7       Josh Wills
8         Leo Duck
Name: name, dtype: object

In [43]:
data[['name']] # otrzymuje obiekt DataFrame

name
0    Allen Smith
1        S Kumar
2    Jack Morgan
3      Ying Chin
4  Dheeraj Patel
5  Satyam Sharma
6   James Authur
7     Josh Wills
8       Leo Duck

### Filtrowanie po rzedach
___
w tym celu moge uzyc indeksow, fragmentow lub warunkow

In [50]:
# filtruje dane za pomoca indeksow
data.filter([0,1,2], axis = 0)

name   age   income gender  department grade  performance_score
0  Allen Smith  45.0      NaN    NaN  Operations    G3                723
1      S Kumar   NaN  16000.0      F     Finance    G0                520
2  Jack Morgan  32.0  35000.0      M     Finance    G2                674

In [58]:
#filtruje dane za pomoca fragmentowania:
data[2:5]

name   age   income gender  department grade  performance_score
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674
3      Ying Chin  45.0  65000.0      F       Sales    G3                556
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711

In [67]:
#filturuje dane pod wzgledem  jednej okreslonej wartosci
data[data.gender == 'F'] 

name   age   income gender  department grade  performance_score
1        S Kumar   NaN  16000.0      F     Finance    G0                520
3      Ying Chin  45.0  65000.0      F       Sales    G3                556
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711
6   James Authur  54.0      NaN      F  Operations    G3                 53
7     Josh Wills  54.0  52000.0      F     Finance    G3                901

In [77]:
# filtruje dane pod wzgledem wiele kolumn za pomoca funkcji isin():
data[data.department.isin(['Sales','Finance'])]

name   age   income gender department grade  performance_score
1        S Kumar   NaN  16000.0      F    Finance    G0                520
2    Jack Morgan  32.0  35000.0      M    Finance    G2                674
3      Ying Chin  45.0  65000.0      F      Sales    G3                556
5  Satyam Sharma   NaN  62000.0    NaN      Sales    G3                649
7     Josh Wills  54.0  52000.0      F    Finance    G3                901
8       Leo Duck  23.0  98000.0      M      Sales    G4                709

In [79]:
#Filtruje pracownikow , ktorzy maja skutecznosc powyzej 700
data[data.performance_score >= 700]

name   age   income gender  department grade  performance_score
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711
7     Josh Wills  54.0  52000.0      F     Finance    G3                901
8       Leo Duck  23.0  98000.0      M       Sales    G4                709

In [84]:
#Filtruje pracownikow , ktorzy maja skutecznosc miedzy 500 a 700
data[(data.performance_score >= 500) & (data.performance_score <=700)]

name   age   income gender department grade  performance_score
1        S Kumar   NaN  16000.0      F    Finance    G0                520
2    Jack Morgan  32.0  35000.0      M    Finance    G2                674
3      Ying Chin  45.0  65000.0      F      Sales    G3                556
5  Satyam Sharma   NaN  62000.0    NaN      Sales    G3                649

# Rozwiazywanie kwestii brakujacych wartosci
___

Moge rozwiazywac ich kwestie na kilka sposobow:
* usuwanie rekordow zawierajacych brakujace wartosci
* reczne uzupelnianie brakujacych wartosci
* uzypelnianie brakujacych wartosciami tendencji centralnej czyli srednia(w cechach numerycznych ), mediana (w cechach porzadkowych) czy dominanta(najszczesciej potwarzajaca wartosc w cechach kategorialnych)
* uzupelnianie najbardziej prawdopodobna wartoscia przy uzyciu ML , takich jak regresja, drzewa decyzyjne , algorytm KKN (K - nearest neighbours)


dla usuwanie brakujacych wartosci uzywam zwykle funkcji `dropna()`, przyjmuje arugment jeden `how`, ktory z kolej przyjmuje wartosci `all`, `any`
* `how = 'all'` - usuwa wzystkie rzedy zawierajace NaN lub brakujace
* `how = 'any'` - usuwa okreslone rzedy zawierajace NaN lub brakujace


In [86]:
data

name   age   income gender  department grade  performance_score
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723
1        S Kumar   NaN  16000.0      F     Finance    G0                520
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674
3      Ying Chin  45.0  65000.0      F       Sales    G3                556
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711
5  Satyam Sharma   NaN  62000.0    NaN       Sales    G3                649
6   James Authur  54.0      NaN      F  Operations    G3                 53
7     Josh Wills  54.0  52000.0      F     Finance    G3                901
8       Leo Duck  23.0  98000.0      M       Sales    G4                709

In [93]:
data.dropna() # usywa wszystkie rzedy z waroscia NaN

name   age   income gender  department grade  performance_score
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674
3      Ying Chin  45.0  65000.0      F       Sales    G3                556
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711
7     Josh Wills  54.0  52000.0      F     Finance    G3                901
8       Leo Duck  23.0  98000.0      M       Sales    G4                709

#### Uzupelnenia brakujacej wartosci
___
`fillna()`- moge uzupelnic brakujace wartosci, przyjmnuje jedna wartosc , ktora bedzie wstawiana w pustych pozycjach zamiast NaN , ewentualnie srednia, mediana, dominanta

In [101]:
# w kolumnie 'age' zamiast wartosci NaN wstawiam srednia wieku pracownikow
data['age'] = data.age.fillna(data.age.mean())
data['age']

0    45.000000
1    40.428571
2    32.000000
3    45.000000
4    30.000000
5    40.428571
6    54.000000
7    54.000000
8    23.000000
Name: age, dtype: float64

In [116]:
#w kolumnie 'income' zamiast NaN wstawiam mediane zarobkow pracownikow 
data['income'] = data.income.fillna(data.income.median())
data

name        age   income gender  department grade  \
0    Allen Smith  45.000000  52000.0    NaN  Operations    G3   
1        S Kumar  40.428571  16000.0      F     Finance    G0   
2    Jack Morgan  32.000000  35000.0      M     Finance    G2   
3      Ying Chin  45.000000  65000.0      F       Sales    G3   
4  Dheeraj Patel  30.000000  42000.0      F  Operations    G2   
5  Satyam Sharma  40.428571  62000.0    NaN       Sales    G3   
6   James Authur  54.000000  52000.0      F  Operations    G3   
7     Josh Wills  54.000000  52000.0      F     Finance    G3   
8       Leo Duck  23.000000  98000.0      M       Sales    G4   

   performance_score  
0                723  
1                520  
2                674  
3                556  
4                711  
5                649  
6                 53  
7                901  
8                709

In [124]:
# kolumnie 'gender' zamiast Nan wstawiam wartosc ktora najszczej wystepuje (dominante)
data['gender'] = data.gender.fillna(data.gender.mode()[0])
data

name        age   income gender  department grade  \
0    Allen Smith  45.000000  52000.0      F  Operations    G3   
1        S Kumar  40.428571  16000.0      F     Finance    G0   
2    Jack Morgan  32.000000  35000.0      M     Finance    G2   
3      Ying Chin  45.000000  65000.0      F       Sales    G3   
4  Dheeraj Patel  30.000000  42000.0      F  Operations    G2   
5  Satyam Sharma  40.428571  62000.0      F       Sales    G3   
6   James Authur  54.000000  52000.0      F  Operations    G3   
7     Josh Wills  54.000000  52000.0      F     Finance    G3   
8       Leo Duck  23.000000  98000.0      M       Sales    G4   

   performance_score  
0                723  
1                520  
2                674  
3                556  
4                711  
5                649  
6                 53  
7                901  
8                709

### Rozwiazywanie kwestii elementow odstajacych 
___
elemetny odstajace dzielimy na **jednowymiarowe i wielowymiarowe**


In [153]:
import pandas as pd

data = pd.read_csv("data/csv/employee.csv")
data

name   age   income gender  department grade  performance_score
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723
1        S Kumar   NaN  16000.0      F     Finance    G0                520
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674
3      Ying Chin  45.0  65000.0      F       Sales    G3                556
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711
5  Satyam Sharma   NaN  62000.0    NaN       Sales    G3                649
6   James Authur  54.0      NaN      F  Operations    G3                 53
7     Josh Wills  54.0  52000.0      F     Finance    G3                901
8       Leo Duck  23.0  98000.0      M       Sales    G4                709

##### Uywaninie elementow odstajacyh  za pomoca sredniej i odchylenia standarowego

In [158]:
#obliczam gorna granice  = srednia + 3 * odchyl. standr.
upper_limit = data['performance_score'].mean() + 3 * data['performance_score'].std()

#obliczam dolna granice  = srednia - 3 * odchyl. standr.
lower_limit = data['performance_score'].mean() - 3 * data['performance_score'].std()
print(f'Dolny limit: {lower_limit}')
print(f'Gorny limit: {upper_limit}')

data = data[(data['performance_score'] < upper_limit) & (data['performance_score'] > lower_limit)]
data

Dolny limit: -96.34906866034851
Gorny limit: 1317.6824019936817


name   age   income gender  department grade  performance_score
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723
1        S Kumar   NaN  16000.0      F     Finance    G0                520
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674
3      Ying Chin  45.0  65000.0      F       Sales    G3                556
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711
5  Satyam Sharma   NaN  62000.0    NaN       Sales    G3                649
6   James Authur  54.0      NaN      F  Operations    G3                 53
7     Josh Wills  54.0  52000.0      F     Finance    G3                901
8       Leo Duck  23.0  98000.0      M       Sales    G4                709

##### Usywanie elementow odstajacyh za pomoca wartosci percentylowych
Jako dolna granice wzialem .01 percentyl , natomiast gorna granice wzilem .99 percentyl. To pozwala wykluczac odstajace elementy w DataFrame

In [4]:
import pandas as pd
data = pd.read_csv('data/csv/employee.csv')
data

name   age   income gender  department grade  performance_score
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723
1        S Kumar   NaN  16000.0      F     Finance    G0                520
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674
3      Ying Chin  45.0  65000.0      F       Sales    G3                556
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711
5  Satyam Sharma   NaN  62000.0    NaN       Sales    G3                649
6   James Authur  54.0      NaN      F  Operations    G3                 53
7     Josh Wills  54.0  52000.0      F     Finance    G3                901
8       Leo Duck  23.0  98000.0      M       Sales    G4                709

In [171]:
upper_limit = data['performance_score'].quantile(.99)
lower_limit = data['performance_score'].quantile(.01)
print(f'Dolny limit: {lower_limit}')
print(f'Gorny limit: {upper_limit}')

data = data[(data['performance_score'] < upper_limit) & (data['performance_score'] > lower_limit)]
data

Dolny limit: 90.36
Gorny limit: 886.76


name   age   income gender  department grade  performance_score
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723
1        S Kumar   NaN  16000.0      F     Finance    G0                520
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674
3      Ying Chin  45.0  65000.0      F       Sales    G3                556
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711
5  Satyam Sharma   NaN  62000.0    NaN       Sales    G3                649
8       Leo Duck  23.0  98000.0      M       Sales    G4                709

### Techniki kodowania cech
___

Modele Ml, sa modelami matematycznymi, dla tego , dla swoich olbiczen potrzebuja wartosci stale, lub zmiennoprzecinkowe. Dla tego musimy przekszalcac cechy kategorialne w numeryczne. Wydajnosc tekiego ML , zalezy od wybranej techniki kodowania.

### Kodownanie  ''goracojedynkowe" (ang. onehotencoder)
___
wartosci kategorialne zostana zatapione wartosciami binarnymi czyli *1* i *0*

In [181]:
# Oczytuje dane
data=pd.read_csv('data/csv/employee.csv')

# Kodowanie 'goracojedynkowe'
encoded_data = pd.get_dummies(data['gender'])

# Laczy zakodowane dane z pierwotna ramka danych
data = data.join(encoded_data)
data.head()

name   age   income gender  department grade  performance_score  \
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723   
1        S Kumar   NaN  16000.0      F     Finance    G0                520   
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674   
3      Ying Chin  45.0  65000.0      F       Sales    G3                556   
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711   

   F  M  
0  0  0  
1  1  0  
2  0  1  
3  1  0  
4  1  0

##### W powyzszym listingu widze ze zostaly dodane dwie kolumny 'F' i 'M', wartosci w ktorej sa jednyki - co oznacza Truem i zera - False

##### Moge uzyskac takie same rezultaty za pomoca klasy `OneHotEncoder` z biblioteki  `scikit-learn`

In [192]:
# imporuje koder "goracojedynkowy"
from sklearn.preprocessing import OneHotEncoder

#tworze obiekt kodera "goracojedynkowy"
onehotencoder = OneHotEncoder()

#wstawiam w kolumne 'gender' zamiat NaN , wartosc dominantna
data['gender'] = data.gender.fillna(data.gender.mode()[0])

#dopasowuje i przekszalcam kolumne gender
onehotencoder.fit_transform(data[['gender']]).toarray()

array([[1., 0.],
       [1., 0.],
       [0., 1.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [0., 1.]])

### Kodowanie etykietowe:
___
Kodowanie etykietowe jest znane jako kodowanie liczbami calkowitymi. 

In [11]:
#importuje klase LabelEncoder
from sklearn.preprocessing import LabelEncoder

#odczytalem dane
data = pd.read_csv("data/csv/employee.csv")

#tworze obietk kodera etykietowego
label_encoder = LabelEncoder()

encoded_data = label_encoder.fit_transform(data['department'])
encoded_data

array([1, 0, 0, 2, 1, 2, 1, 0, 2])

##### W powyzszym przykladzie przeprowadzilem proste kodowanie etykietowe,  1 - Operations,  0 - Finance, 2 - Sales
`fit_transform(data['department'])`- sluzy do dopasowania i przekrztalcenia kolumny na cyfry

##### Teraz zakoduje kolumne department za pomoca klasy LabelEndocder, wykonuje odwotna transormacje zakodowanych etykiet:
`inverse_transform()`- sluzy do odwrotnej transformacji zakodowanych etykiet

In [204]:
inverse_encode = label_encoder.inverse_transform([0,1,2,0,2])
print(inverse_encode)

['Finance' 'Operations' 'Sales' 'Finance' 'Sales']


### Koder zmiennych porzadkowych
___
Kodowanie zmiennych porzadkowych jest podobne do kodowania etykietowego , ale musimy dodatkowo uwzgledniac kolejnosc kodowanych elementow. Mozemy zdefiniowac kolejnosc wartosci jako lise i przekazac ja jako parapemert do klasy

In [238]:
from sklearn.preprocessing import OrdinalEncoder

order_encoder=OrdinalEncoder(categories=['G0','G1','G2','G3','G4'])
data['grade_encode'] = label_encoder.fit_transform(data['grade'])
data

name   age   income gender  department grade  performance_score  \
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723   
1        S Kumar   NaN  16000.0      F     Finance    G0                520   
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674   
3      Ying Chin  45.0  65000.0      F       Sales    G3                556   
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711   
5  Satyam Sharma   NaN  62000.0    NaN       Sales    G3                649   
6   James Authur  54.0      NaN      F  Operations    G3                 53   
7     Josh Wills  54.0  52000.0      F     Finance    G3                901   
8       Leo Duck  23.0  98000.0      M       Sales    G4                709   

   grade_encoded  grade_encode  
0              2             2  
1              0             0  
2              1             1  
3              2             2  
4              1             1  
5              2             2  
6              2             2  
7              2             2  
8              3             3

### Metody skalowanie cech
___
Porownywanie cech mieszczace sie o roznych skalach, np wiek opisujemy od 0 do 100, a wynagrodzenie od 0 do 100000 w danej walucie. Cechy w ktorych wielkosci sa bardzo duze , beda mialy w ML wieksza wage,dla tego mozemy taki problem rozwiazac za pomoca **skalowaniu cech lub normalizowaniu cech.**

Metody skalowanie cech:
> * Skalowanie standordowe lub standaryzacja - obliczane za pomoca std i srednej 
> * Skalowanie min-max
> * Skalowanie odporne

In [9]:
#importuje klase StandardScaler
from sklearn.preprocessing import StandardScaler

#inicjalizuje klase StandardScaler
scaler = StandardScaler()

#skaluje dane
scaler.fit(data['performance_score'].values.reshape(-1,1)) 
data['performance_std_scaler']=scaler.transform(data['performance_score'].values.reshape(-1,1))
data.head()

name   age   income gender  department grade  performance_score  \
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723   
1        S Kumar   NaN  16000.0      F     Finance    G0                520   
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674   
3      Ying Chin  45.0  65000.0      F       Sales    G3                556   
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711   

   performance_std_scaler  
0                0.505565  
1               -0.408053  
2                0.285037  
3               -0.246032  
4                0.451558

### Przeksztalcenia cech
___
za pomoca **dyskretyzacji** mozemy przeksztalcac kolumne. Na przyklad przepisywanie zamiast wieku pracownikow , etykiety np. mlody, dorosly, stary

In [14]:
data = pd.read_csv("data/csv/employee.csv")

def performance_grades(score):
    if score >= 700:
        return 'A'
    if score < 700 and score >= 500:
        return 'B'
    else:
        return 'C'
    
data['performance_grade'] = data["performance_score"].apply(performance_grades)
data

name   age   income gender  department grade  performance_score  \
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723   
1        S Kumar   NaN  16000.0      F     Finance    G0                520   
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674   
3      Ying Chin  45.0  65000.0      F       Sales    G3                556   
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711   
5  Satyam Sharma   NaN  62000.0    NaN       Sales    G3                649   
6   James Authur  54.0      NaN      F  Operations    G3                 53   
7     Josh Wills  54.0  52000.0      F     Finance    G3                901   
8       Leo Duck  23.0  98000.0      M       Sales    G4                709   

  performance_grade  
0                 A  
1                 B  
2                 B  
3                 B  
4                 A  
5                 B  
6                 C  
7                 A  
8                 A

### Rodzielanie cech
___
za pomoca rodzieleniu cech mozemy dzielic np. cechy danych osobowych na pierwsze imie, drugie imie, nazwisko, a kolumne adresu na misto , ulice, nr budunku, nr mieszkanie i tak dalej.

In [23]:
data = pd.read_csv("data/csv/employee.csv")

data['first_name'] = data['name'].str.split(" ").map(lambda var: var[0])
data['last_name'] = data['name'].str.split(" ").map(lambda var: var[1])
data.head()

name   age   income gender  department grade  performance_score  \
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723   
1        S Kumar   NaN  16000.0      F     Finance    G0                520   
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674   
3      Ying Chin  45.0  65000.0      F       Sales    G3                556   
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711   

  first_name last_name  
0      Allen     Smith  
1          S     Kumar  
2       Jack    Morgan  
3       Ying      Chin  
4    Dheeraj     Patel